Upgrade to Rails 4

erino opened the pull request.

erino committed:
rails ~> 4.2 and related gems

* devise ~> 3.5
* mongoid ~> 4.0
* mongoid-paperclip ~> 0

erino committed:
Update devise config for new version

erino committed:
actionpack-page_caching ~> 1.1

caches_page has been extracted from Rails core in 4.0

erino committed:
Allow dynamic attributes

The previous Mongoid version defaulted to use dynamic attributes which
Slices did not override. Including `Mongoid::Attributes::Dynamic`
preserves this functionality.

erino committed:
Update Paperlip attachment validation to v4 syntax

https://github.com/thoughtbot/paperclip/tree/v4.3.2#validations

erino committed:
Remove Moped namespace from BSON::ObjectId

Mongoid BSON implementation has been removed in favor of the 10gen
BSON gem. The API is consistant with the Moped version.

erino committed:
Update specs with new Rails form helper attributes

erino committed:
Add _id to minimal scope

The id field isn't automatically included, so we need
to explicitly include it. This is fixed in Mongoid 5

https://github.com/mongodb/mongoid/commit/d873923910f5a4a419bc5a4fd4e9c7adc317e29b

erino committed:
Update spec to reflect new pluralization rules for new models

erino committed:
Update dummy rails app configuration

Using a combination of:
  rake rails:update
  rails new dummy -OT

erino committed:
Add strong parameters to controllers

http://guides.rubyonrails.org/action_controller_overview.html#strong-parameters

erino committed:
Remove depreciated string terminator from define_callbacks

String based terminators were removed in 4.0
https://github.com/rails/rails/commit/b342d2baef4c0aa108f62e67d2bdb9b6641d6363

erino committed:
Ensure BSON Object ids are serialized as strings

With Mongoid 4 the JSON serialization for an Object Id
is: `{"$oid": "...."}`. This patch ensures that the ids are serialized
as strings which is the Mongoid 3 behaviour.

erino committed:
Refactor Admin#as_json

erino committed:
Use #Admin.super_user? rather then #super?

Admin#super? was only used to genereate a json object with the key
`super_user`, we now have a consistant use of `super_user`

erino committed:
Mock #as_json in page double

This was causing a stack level too deep error.

erino committed:
Remove commented code

erino committed:
Remove save call from Asset#reprocess_for

Paperclip#reprocess! calls save, so we don't
need to call it again

erino committed:
mongoid-observers ~> 0

Observers have been removed from Mongoid 4

erino committed:
Ensure Page#slice_attachment_asset_ids is free from duplicates

We're only concerned with which assets are used on a page, not
how many times an asset is used.

erino committed:
Tidy up serialization of Page#author

erino committed:
Ignore uploaded and generated files in the dummy app

erino committed:
Rename text_search to basic_text_search

In Mongoid 4.0 text_search is used to access the MongoDB text search
feature. Renamining the current text_search to basic_text_serach avoids
conflicts.

http://www.rubydoc.info/gems/mongoid/4.0.0/Mongoid%2FContextual%2FMongo%3Atext_search

erino committed:
Check that there are pages before writing the page cache

If the asset orginally has no pages and is then added to a page
then saved #pages is still empty. Calling pages.exists? forces a
refresh

erino committed:
Use have_xpath to spec only the attributes we're concerned with

Rails 4 has changed the HTML attributes generated in the image_tag,
so rather then changing the `expected_html`, we now only spec the
attributes we're concerned with.

erino committed:
Add document_root_element to api example group

Rails::Dom::Testing now implictily selects from
the document_root_element.

https://github.com/rails/rails-dom-testing/blob/master/README.md#selector-assertions

erino committed:
rspec ~> 3.4

erino committed:
Fix Rails asset path for upload icons

* Remove duplication in how icon paths are stored
* Remove extra file name extension added to path

erino committed:
capybara ~> 2.5

erino committed:
Provide default values for localized fields

If there's no default, fields can end up in an invalid state.
https://jira.mongodb.org/browse/MONGOID-4054

erino committed:
Use a different DB for localized development

Mongoid doesn't support switching from localized to non-localized
fields, so switching between states causes the app to break. This
makes local development easier as the database is automatically
changed when localization is enabled or disabled.

erino committed:
Make another change on a new page to trigger an update

erino committed:
redcarpet ~> 3.3

erino committed:
RedCloth ~> 4.2

erino committed:
poltergeist ~> 1.7

calumgunn commented:
Looks good to me!

bensymonds commented:
@erino There are quite a lot of things in here where it's not really obvious why they were done. Would it be too much faff to add a description (one or two lines) to the commit messages? It feels like it would be good to have this stuff documented so if/when it causes problems we know the background.

erino commented:
@bensymonds I've added a few descriptions, and squashed a few commits. Are there any other commits that you think need expanding?

erino committed:
A start on Slices 3 update guide

bensymonds commented on app/models/attachment.rb:
```diff
@@ -7,7 +8,11 @@ class Attachment
 
   def as_json *args
-    result = attributes.symbolize_keys.merge asset: asset.as_json
+    result = attributes.symbolize_keys.except(:_id, :_type, :asset_id).merge(
```
Is it ok that `_id` is changing to `id`, and `_type` is being removed?

bensymonds replied:
I think I can see the change in the thing that consumes this ([here](https://github.com/withassociates/slices/pull/116/files#diff-771bae8f8b744e9e061f0b83820792b9R93), so I guess it's ok. But this came in as part of making sure they were all strings, so wasn't sure why the name had to change too. And there is a [spec](https://github.com/withassociates/slices/blob/master/spec/models/slice/write_attributes_spec.rb#L39) which looks like it needs updating..? Except it's not failing, so maybe I'm missing something.

erino replied:
It is ok, but i'm going to change the back to `_id` to make it consistant with the rest of the API

bensymonds replied:
:+1:

bensymonds commented on app/models/admin.rb:
```diff
@@ -31,18 +31,17 @@ class Admin
-    end
+    current_admin_id = options[:current_admin].try(:id)
+    {
+      _id: id.to_s,
```
Should this change to `id`? It seems like we are trying to switch to that, right? (https://github.com/withassociates/slices/pull/116/files#r50683732)

erino replied:
It should stay as `_id` as that what (most) of the API uses

bensymonds replied:
Ok fair enough. I probably would choose `id` over `_id` but for now `_id` is fine - stay consistent. Then we can change them all later if we want to.

bensymonds commented on Gemfile:
```diff
@@ -2,15 +2,19 @@ source 'https://rubygems.org'
-gem 'test-unit', platforms: [:ruby_22]
 
-group :development do
+group :development, :test do
```
The removal of the `:test group was deliberate - happened in https://github.com/withassociates/slices/commit/d22e903db0ea96f63df80fd58038509033c0fe49. How come you added it back? Was it necessary for upgrading to rails 4 (it was in that commit)?

I think the reasoning at the time was that as Slices is a gem and not an app it never runs in a "production" mode - the gems in here are for testing and development only. Development gems are superset of test gems, so made sense to make "test" be the _default_ group.

erino replied:
That change must have been lost in a rebase. I think we should revisit `d22e903db0ea96f63df80fd58038509033c0fe49` as I don't see the reason for including test gems in all environments.

bensymonds replied:
But that's my point - what other environments are there? There is no environment where you don't want the test gems, is there?

erino replied:
What about development?, the dummy app uses the same Gemfile: https://github.com/withassociates/slices/blob/rails-4/spec/dummy/Gemfile

bensymonds replied:
I don't know what you mean about the dummy app - it has its own Gemfile (the one you linked) and Gemfile.lock, so it's unaffected by anything going on in the main Slices versions of those files, isn't it? So none of the test-y gems are loaded when that guy runs.

bensymonds commented on slices.gemspec:
```diff
@@ -13,15 +13,14 @@ Gem::Specification.new do |s|
-  s.add_dependency 'rails'             , '~> 3.2.18'
-  s.add_dependency 'redcarpet'         , '~> 3.0.0'
-  s.add_dependency 'RedCloth'          , '~> 4.2.9'
+  s.add_dependency 'mongoid'           , '> 4.0.0', '< 5.0.0'
```
This is just `~> 4.0`, isn't it? Shall we write it like that to keep consistent with the other ones?

bensymonds commented on app/controllers/pages_controller.rb:
```diff
@@ -1,7 +1,5 @@
 class PagesController < SlicesController
 
-  caches_page :virtual_error_pages
```
What does this mean for error pages then? Are they no longer cached?

bensymonds replied:
(Maybe something to add to the upgrade guide, if the idea is that the app embedding slices has to do it itself now.)

bensymonds commented on app/controllers/slices_controller.rb:
```diff
@@ -9,7 +9,7 @@ class SlicesController < ActionController::Base
 
   around_filter :set_locale
-  define_callbacks :render_page, terminator: "response_body"
+  define_callbacks :render_page
```
Why this (commit doesn't say)?

erino replied:
String based terminators have been depreciated, I've added some info in the commit

bensymonds replied:
Ok they were deprecated. But why does that mean this can be _removed_? (I have never figured out what this line does.)

bensymonds commented on lib/slices/tree.rb:
```diff
@@ -39,7 +39,7 @@ module Tree
       field :show_in_nav, type: Boolean, default: false
 
-      scope :minimal, ->{ only(%w(_type page_id path external_url name has_content)) }
+      scope :minimal, ->{ only(%w(_id _type page_id path external_url name has_content)) }
```
Commits says

> We need ids to generate links

Can you explain a bit more? Interested to know how it worked before now.

erino replied:
I'm guessing that the previous version of Mongoid automatically included id when using `only`, now it's possible to not include `_id` which breaks `link_to` as there is no id

erino replied:
It's fixed in Mongoid 5 https://github.com/mongodb/mongoid/commit/d873923910f5a4a419bc5a4fd4e9c7adc317e29b

bensymonds commented on spec/controllers/admin/pages/find_all_slices_spec.rb:
```diff
@@ -12,11 +12,11 @@
 
     let :page do
-      double(:page).as_null_object
+      double(:page, as_json: {}).as_null_object
```
Commit says:

> This was causing a stack level too deep error.

But it would be good to know why _that_ was happening. What's the root cause?

erino replied:
I think it was a Rspec double continually calling `as_json` on its self

bensymonds commented on spec/lib/slices/has_slices/text_search_in_spec.rb:
```diff
@@ -33,18 +33,22 @@ class DummyPage < Page
 
+      it "search contents of String fields" do
+        expect(DummyPage.basic_text_search("Hello")).to be_a(Mongoid::Criteria)
+      end
```
Whacky indentation.

bensymonds commented on spec/lib/slices/has_slices/text_search_in_spec.rb:
```diff
@@ -33,18 +33,22 @@ class DummyPage < Page
       end
     end
 
+      it "search contents of String fields" do
```
Whacky description.

bensymonds commented on spec/deprecated/functional_page_controller_spec.rb:
```diff
@@ -13,12 +13,12 @@
     it "sets the page type if set" do
       get new_admin_page_path(parent_id: @page.to_param, type: 'article')
-      assert_select 'input[type=hidden][name=type][value=article]'
+      assert_select html_document.root, 'input[type=hidden][name=type][value=article]'
```
Why was this necessary?

bensymonds commented on app/models/page.rb:
```diff
@@ -15,15 +16,15 @@ class Page
-  field :meta_description, localize: Slices::Config.i18n?
-  field :title, localize: Slices::Config.i18n?
+  field :meta_description, localize: Slices::Config.i18n?, default: ''
+  field :title, localize: Slices::Config.i18n?, default: ''
```
> If there's no default, fields can end up in an invalid state.

Sounds interesting, tell us more!

erino replied:
I started to look into this, but I didn't find anything conclusive. The follow spec fails if the above defaults are missing:

``` ruby
rspec spec/apis/pages/update_spec.rb:133
```

bensymonds replied:
https://jira.mongodb.org/browse/MONGOID-4054. I think add that link as a comment somewhere here so reader knows why the defaults are being set. Then eventually it'll be fixed and released in Mongoid and this change can be reverted.

erino replied:
I've added a link to the bug report in the commit and updated the guides to say a default should be set.

bensymonds commented on spec/requests/admin/page/create_page_spec.rb:
```diff
@@ -35,6 +35,7 @@
     context "saving the page" do
       before do
         fill_in 'Page Name', with: 'New Page'
+        fill_in 'Meta Description', with: 'new meta description'
```
> Make another change on a new page to trigger an update

I don't understand why this is necessary.

bensymonds commented on app/helpers/admin/site_maps_helper.rb:
```diff
@@ -7,7 +7,7 @@ module Admin::SiteMapsHelper
   # @!visibility private
   def sitemap_pages
-    @sitemap_pages ||= proper_pages.only(:id, :name, :page_id, :path).group_by(&:page_id)
+    @sitemap_pages ||= proper_pages.only(:id, :name, :page_id, :path, :position, :role).group_by(&:page_id)
```
As with https://github.com/withassociates/slices/pull/116/files#r50687277, I don't understand why this became necessary.

erino replied:
As per the other `only` scope, this fields are needed to render the site map. If they're not specified here then it's not possible to order pages or draw a special icon depending on the role.

bensymonds replied:
Ok but with that other `only` scope the theory is that older Mongoid used to include `_id` by default. That wouldn't be the case with these fields, so something else must have changed...

erino replied:
Yep #106

bensymonds replied:
Ah ok, yeah I forgot this was changed recently. So does that mean `master` is currently broken? In which case should we merge this fix first, before rails 4, and do a cheeky point release? Otherwise all the sites we're going to leave on rails 3 will pick this up and be broken.

erino replied:
I've not checked, but that could well be the case!

bensymonds replied:
I have checked, and it does indeed seem to be broken.

bensymonds replied:
https://github.com/withassociates/slices/pull/152. Then this can be rebased.

erino committed:
Update mongoid config for new version

calumgunn commented:
I've got a commit here that fixes a bunch of typos in documentation comments and spec names etc! Shall I push it to this branch, or a new PR branched off of this one?

bensymonds commented on guides/internationalization.md:
```diff
@@ -6,11 +6,12 @@ The first step is to read and follow the [Rails I18n Guide](http://guides.rubyon
 
-Firstly add `localize: true` to all text fields which need localzing.
+Firstly add `localize: true` to all text fields which need localizing, make sure you add a 
+default value as there is a bug in Mongoid which means the value may not be set correctly.
```
Probably good to put a link to the bug here too (as well as commit message). That way users can see for themselves if they need to do it (it might get fixed in future 4.x release).

bensymonds commented:
@calumgunn If it's not rails 4 specific then why not put it into master? If we have the opportunity to make another, better v2.x release then we may as well take it. Plus it's nicer to not add more to this PR if we can help it - it's large enough as it is.

erino committed:
Paperclip 4.3.3 introduces a dependancy on ActiveRecord

We'll exclude version 4.3.3, and this will hopefully be fixed in 4.3.4
https://github.com/thoughtbot/paperclip/commit/2e6335532ac24012cf473b60994d09f14952b891#diff-8122279dfef186be2dd5cf7fddf1b498R14
